### PR TITLE
Add Circle Loss Function for Similarity/Metric Learning Tasks.

### PR DESCRIPTION
**Pull request by ma7555**
**Overview**

This PR adds the implementation of the Circle Loss function to Keras, addressing issue [#20421](https://github.com/keras-team/keras/issues/20421). Circle Loss is designed for metric learning tasks, aiming to minimize within-class distances and maximize between-class distances in the embedding space.

1. **Normal Within-Batch Negative/Positive Pairs mining**
   - Directly use the Circle class as the case with all other losses.
```python
import keras

(x_train, y_train), (x_test, y_test) = keras.datasets.cifar100.load_data()

model = keras.Sequential()
model.add(keras.layers.InputLayer(shape=(32, 32, 3)))
model.add(keras.layers.Rescaling(1.0 / 255, offset=-1))
for i in range(3):
    model.add(
        keras.layers.Conv2D(
            32, (3, 3), padding="valid", activation="relu", name=f"conv_{i}"
        )
    )
    model.add(keras.layers.MaxPooling2D((2, 2)))
model.add(keras.layers.Flatten())
model.add(keras.layers.Dense(64, activation=None))
model.add(keras.layers.UnitNormalization())

model.compile(
    optimizer="adam",
    loss=keras.src.losses.losses.Circle(),
)
model.fit(x_train, y_train, validation_data=(x_test, y_test), epochs=5, batch_size=2048)
```
```
Epoch 1/5
25/25 ━━━━━━━━━━━━━━━━━━━━ 10s 362ms/step - loss: 54.0521 - val_loss: 47.4749
Epoch 2/5
25/25 ━━━━━━━━━━━━━━━━━━━━ 9s 370ms/step - loss: 46.7501 - val_loss: 45.2312
Epoch 3/5
25/25 ━━━━━━━━━━━━━━━━━━━━ 9s 365ms/step - loss: 44.8289 - val_loss: 44.0905
Epoch 4/5
25/25 ━━━━━━━━━━━━━━━━━━━━ 9s 362ms/step - loss: 43.7766 - val_loss: 43.6335
Epoch 5/5
25/25 ━━━━━━━━━━━━━━━━━━━━ 9s 347ms/step - loss: 43.2757 - val_loss: 43.0075
```

2. **Support for Cross-Batch Memory (XBM) Training**
   - Introduces `ref_embeddings` and `ref_labels` parameters to the loss function which defaults to y_pred and y_true if not passed.
   - Allows the use of embeddings and labels from a memory bank or previous batches.
   - I wrote an example of usage with [XBM](https://gist.github.com/ma7555/2f56163ce815e042e018f62e16f7c2c8)

3. **Fixes to `logsumexp` Function for Torch/JAX**

   - keras.ops.logsumexp is numerically unstable in JAX and PyTorch backends when an entire row contains `-inf` values (e.g., when a sample has no positive or negative pairs), resulting in NaN outputs instead of the expected `-inf`. In native Jax/Torch, this operation returns the expected `-inf` but keras.ops does not use the native backend implementation. While you may have a valid reason for this that i am not aware of, I have modified the ops in the last commit to use the native operation since all tests pass. The test fails on numpy backend as it is left as it is waiting for the reviewer comment on which path to follow. numpy backend can be replaced with `scipy.special.logsumexp`

**Testing**

- Correctness testing against pytorch implementation can be reviewed in [colab](https://colab.research.google.com/drive/1HrxUPmhJY2rCmD3cL1TJpmAt-5JFiQGn?usp=sharing)

Notes:
* This is a draft PR, I will still test the loss function in real large model training scenario after the draft is approved.
* This loss function requires training on L2 normalized embeddings (like [CosineSimilarity](https://github.com/keras-team/keras/blob/272bb901dd207a57656286284cee1aa8c28061cb/keras/src/losses/losses.py#L218)). Cosine Similarity handles this internally, but I did not follow this and added in the docs that the last layer should be [UnitNormalization](https://github.com/keras-team/keras/blob/272bb901dd207a57656286284cee1aa8c28061cb/keras/src/layers/normalization/unit_normalization.py#L7)
* `remove_diagonal` argument is kept as an argument in cases for advanced users to train across completely different memory bank. Like two different datasets, one used as query and the other as reference. In all other cases, this should be always `True`

### CODE REVIEW

File: keras/src/losses/losses.py [hunk 1]
```diff
@@ -1403,6 +1404,94 @@ def get_config(self):
         return config
 
 
+@keras_export("keras.losses.Circle")
+class Circle(LossFunctionWrapper):
+    """Computes Circle Loss, a metric learning loss designed to minimize
```
Comment (fchollet) [inline keras/src/losses/losses.py:1409]: Split the description into a one-line summary and a longer description:

```
"""This is the one line summary.

This is the longer description.
...
"""
```

File: keras/src/losses/losses.py [hunk 2]
```diff
@@ -1403,6 +1404,94 @@ def get_config(self):
...
+        keras.layers.Flatten(),
+        keras.layers.Dense(64, activation=None),  # No activation
+        keras.layers.UnitNormalization()  # L2 normalization
+    ])
+
+    model.compile(optimizer="adam", loss=losses.Circle()
+    ```
+
+    Reference:
+        [Yifan Sun et al., 2020](https://arxiv.org/abs/2002.10857)
```
Comment (fchollet) [inline keras/src/losses/losses.py:1458]: Use a list

File: keras/src/losses/losses.py [hunk 3]
```diff
@@ -2406,3 +2495,91 @@ def tversky(y_true, y_pred, alpha=0.5, beta=0.5):
...
+def circle(
+    y_true,
+    y_pred,
+    ref_labels=None,
+    ref_embeddings=None,
+    remove_diagonal=True,
+    gamma=80,
+    margin=0.4,
+):
+    """Computes the Circle loss
```
Comment (fchollet) [inline keras/src/losses/losses.py:2510]: Fix nits in docstring (e.g. missing period here, indentation...)

File: keras/src/losses/losses.py [hunk 4]
```diff
@@ -2406,3 +2495,91 @@ def tversky(y_true, y_pred, alpha=0.5, beta=0.5):
...
+        ops.where(positive_mask, pos_wdists, float("-inf")),
+        axis=1,
+    )
+    n_loss = ops.logsumexp(
+        ops.where(negative_mask, neg_wdists, float("-inf")),
+        axis=1,
+    )
+
+    circle_loss = ops.softplus(p_loss + n_loss)
+    backend.set_keras_mask(circle_loss, circle_loss > 0)
```
Comment (fchollet) [inline keras/src/losses/losses.py:2584]: Depending on how the loss is used, the mask might not be taken into account. How critical is it?

**Review by fchollet — COMMENTED** [review]
Thanks for the PR! Please take a look at the numpy test failure.

File: keras/src/losses/losses.py (hunk 4 above)
Comment (ma7555) [reply keras/src/losses/losses.py:2584]: This masking behaviour is used to mask samples from the batch that has no negative/positive pairs (a solo class with nothing to compare to). When this happens, you want to eleminate it from the loss as it has a loss value of 0 and not masking it can affect the `sum_over_batch_size` reduction (make the loss lower than reality). In pytorch, they use the [AvgNonZeroReducer](https://kevinmusgrave.github.io/pytorch-metric-learning/reducers/#avgnonzeroreducer) by taking the mean only values that is not a zero.

I won't say it is critical because:
  1. Using a data sampler for pair generation like [TFDataSampler](https://github.com/tensorflow/similarity/blob/dc506d63d05d012d3496041c25c1c54981d39565/tensorflow_similarity/samplers/tfdata_sampler.py#L135) or any similar sampler solves this (which is a standard in metric learning).
  2. Using a larger batch size.
  3. Not an issue if it happens every now and then.
  4. If it happens all the time, it means that the data feeding sampler is the problem not the loss function itself.

**Review by fchollet — COMMENTED** [review]
Thanks for the update! To fix the code format check, run `sh shell/format.sh`. You may need to shorten text lines by hand in the docstrings.

**Comment by ma7555** [comment]
@fchollet looks good now, please hold till I test it in real model training scenario

**Comment by ma7555** [comment]
tested and looks fine. [Kaggle Notebook](https://www.kaggle.com/code/ma7555/keras-circleloss-test)
Ready to pull.

**Review by fchollet — APPROVED** [review]
LGTM, thank you for the contribution!
